tests/test_recombination_map.py

```diff
@@ -0,0 +1,437 @@
+#
+# Copyright (C) 2018 University of Oxford
+#
+# This file is part of msprime.
+#
+# msprime is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# msprime is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with msprime.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Tests for the recombination map functionality, mapping continuous physical
+coordinates to discrete genetic loci and vice versa.
+"""
+from __future__ import print_function
+from __future__ import division
+
+import unittest
+import random
+import tempfile
+import os
+import gzip
+
+import numpy as np
+
+import msprime
+
+
+class PythonRecombinationMap(object):
+    """
+    A Python implementation of the RecombinationMap interface.
+    """
+    def __init__(self, positions, rates, num_loci):
+        assert len(positions) == len(rates)
+        assert len(positions) >= 2
+        assert sorted(positions) == positions
+        assert positions[0] == 0
+        # assert positions[-1] == 1
+        self._positions = positions
+        self._sequence_length = positions[-1]
+        self._rates = rates
+        self._num_loci = num_loci
+
+    def get_total_recombination_rate(self):
+        """
+        Returns the effective recombination rate for this genetic map.
+        This is the weighted mean of the rates across all intervals.
+        """
+        x = self._positions
+        effective_rate = 0
+        for j in range(len(x) - 1):
+            length = (x[j + 1] - x[j])
+            effective_rate += self._rates[j] * length
+        return effective_rate
+
+    def _physical_to_genetic_zero_rate(self, x):
+        """
+        If we have a zero recombination rate throughout, then we only have
+        two possible values. Any value < L maps to 0, as this is the start
+        of the interval. If x = L, then we map to num_loci because the interval
+        is half-open. This is consistent with the behaviour we get when we
+        have an interval of zero recombination within the map: any value
+        within the interval is mapped down to the start coordinate, and the
+        endpoint signifies the beginning of the next interval.
```

Comment:
I think this is the right behavior but I don't understand what is meant by "the endpoint signifies the beginning of the next interval" - what interval? I think that an entire physical region of zero recombination, within the map, gets mapped to a single genetic point, including both endpoints.